spec/forms/hub/source_params_form_spec.rb

```diff
@@ -0,0 +1,53 @@
+require "rails_helper"
+
+describe Hub::SourceParamsForm do
+  subject { described_class.new(vita_partner, params) }
+
+  let(:vita_partner) { create :vita_partner }
+  let(:code) { "my_code" }
+  let(:params) do
+    {
+        code: code
+    }
+  end
+  describe "#valid?" do
+    context "when the source parameter is already in use" do
+      context "when the source parameter is" do
+        before do
+          create :source_parameter, vita_partner: vita_partner, code: code
+        end
+
+        it "is invalid with appropriate message" do
+          expect(subject.valid?).to eq false
+          expect(subject.errors[:code]).to include "my_code is already applied to this partner."
+        end
+      end
+
+      context "when it belongs to a different vita partner" do
+        let(:other_vita_partner) { create :vita_partner, name: "Oregano Org" }
+        before do
+          create :source_parameter, vita_partner: other_vita_partner, code: code
+        end
+
+        it "is invalid with appropriate message" do
+          expect(subject.valid?).to eq false
+          expect(subject.errors[:code]).to include "my_code is already in use by <a href=/en/hub/organizations/#{other_vita_partner.id}/edit#source-params-form>Oregano Org</a>."
+        end
+      end
+    end
+
+    context "when the source parameter is unique and valid" do
+      it "is valid" do
+        expect(subject.valid?).to eq true
+      end
+    end
+  end
+
+  describe "#save!" do
+    it "creates a new source parameter for the vita partner" do
```

Comment:
I think it would be good to check the contents of the new source parameter in this test.